internal/controller/grafanaorganization_controller.go

```diff
@@ -39,24 +44,91 @@ type GrafanaOrganizationReconciler struct {
 
 // Reconcile is part of the main Kubernetes reconciliation loop which aims to
 // move the current state of the GrafanaOrganization closer to the desired state.
-// TODO(zirko): Modify the Reconcile function to compare the state specified by
-// the GrafanaOrganization object against the actual organization state, and then
-// perform operations to make the organization state reflect the state specified by
-// the user.
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.0/pkg/reconcile
 func (r *GrafanaOrganizationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
 
-	// TODO(zirko): your logic here
+	logger.Info("Started reconciling Grafana Organization")
+	defer logger.Info("Finished reconciling Grafana Organization")
+
+	grafanaOrganization := &v1alpha1.GrafanaOrganization{}
+	err := r.Client.Get(ctx, req.NamespacedName, grafanaOrganization)
+	if err != nil {
+		return ctrl.Result{}, errors.WithStack(client.IgnoreNotFound(err))
+	}
+
+	logger.WithValues("grafanaOrganization", grafanaOrganization.ObjectMeta.Name)
+
+	// Generate Grafana client
+	grafanaAPI, err := grafanaClient.GenerateGrafanaClient(ctx, r.Client, logger)
```

Comment:
I may be wrong, but it looks like we could create this client once during the bootstrap of the controller, instead of creating it on every reconciliation loop. Both the admin secret and the TLS secret seem to be independent of the reconciled organization. That way we would save these calls to the k8s api on every reconciliation loop.

Then the reconciler would receive the grafana client when creating it.

Reply:
The issue is that when the grafana secret changes then we would have to restart the operator right?

Reply:
Whenever there is a leak :D

Reply:
![](https://media1.tenor.com/m/ZAHC50FvopkAAAAC/futurama-fry.gif)

Reply:
Done

Reply:
This is not done.

BTW, thinking about this, instead of fetching it using the client, why don't we specify the secrets in the `Deployment` and let k8s do the wiring?

Reply:
It's done locally on my machine and with the vs code extension I edited the comment before pushing as it's easier to track. It's done in the latest commit

Reply:
I' afraid of the risk for new installations if the operator comes up before the grafana app but that's a good question

Reply:
It's done, i'm testing it now